File: src/main/java/com/gtladd/gtladditions/common/machine/trait/SuperNotifiableFluidTank.java
package com.gtladd.gtladditions.common.machine.trait;

import com.gregtechceu.gtceu.api.capability.recipe.IO;
import com.gregtechceu.gtceu.api.machine.MetaMachine;
import com.gregtechceu.gtceu.api.machine.trait.NotifiableFluidTank;

import com.lowdragmc.lowdraglib.side.fluid.FluidStack;
import com.lowdragmc.lowdraglib.side.fluid.IFluidTransfer;
import com.lowdragmc.lowdraglib.side.fluid.forge.FluidHelperImpl;

import net.minecraft.core.BlockPos;
import net.minecraft.core.Direction;
import net.minecraft.world.level.Level;
import net.minecraft.world.level.block.entity.BlockEntity;
import net.minecraft.world.level.block.state.BlockState;
import net.minecraftforge.common.capabilities.ForgeCapabilities;
import net.minecraftforge.fluids.capability.IFluidHandler;

import org.jetbrains.annotations.Nullable;

import java.util.function.Predicate;

public class SuperNotifiableFluidTank extends NotifiableFluidTank {

    public SuperNotifiableFluidTank(MetaMachine machine, int slots, long capacity, IO io) {
        super(machine, slots, capacity, io);
    }

    @Override
    public void exportToNearby(Direction... facings) {
        if (!this.isEmpty()) {
            Level level = this.getMachine().getLevel();
            BlockPos pos = this.getMachine().getPos();

            for (Direction facing : facings) {
                exportToTarget(this, this.getMachine().getFluidCapFilter(facing), level, pos.relative(facing), facing.getOpposite());
            }

        }
    }

    public static void exportToTarget(IFluidTransfer source, Predicate<FluidStack> filter, Level level, BlockPos pos, @Nullable Direction direction) {
        BlockState state = level.getBlockState(pos);
        if (state.hasBlockEntity()) {
            BlockEntity blockEntity = level.getBlockEntity(pos);
            if (blockEntity != null) {
                var cap = blockEntity.getCapability(ForgeCapabilities.FLUID_HANDLER, direction).resolve();
                if (cap.isPresent()) {
                    var target = cap.get();
                    for (int srcIndex = 0; srcIndex < source.getTanks(); srcIndex++) {
                        var currentFluid = source.getFluidInTank(srcIndex);
                        if (currentFluid.isEmpty() || !filter.test(currentFluid)) {
                            continue;
                        }

                        var toDrain = currentFluid.copy();
                        var remainAmount = currentFluid.getAmount();
                        do {
                            var filled = target.fill(FluidHelperImpl.toFluidStack(source.drain(toDrain, true)), IFluidHandler.FluidAction.SIMULATE);
                            if (filled > 0) {
                                toDrain = currentFluid.copy();
                                toDrain.setAmount(filled);
                                target.fill(FluidHelperImpl.toFluidStack(source.drain(toDrain, false)), IFluidHandler.FluidAction.EXECUTE);
                                remainAmount -= filled;
                                if (remainAmount <= 0) break;
                            } else {
                                break;
                            }
                            toDrain = currentFluid.copy();
                            toDrain.setAmount(remainAmount);
                        } while (!toDrain.isEmpty());
                    }
                }
            }
        }
    }
}

File: src/main/java/com/gtladd/gtladditions/common/machine/trait/SuperNotifiableFluidTank.kt
package com.gtladd.gtladditions.common.machine.trait

import com.gregtechceu.gtceu.api.capability.recipe.IO
import com.gregtechceu.gtceu.api.machine.MetaMachine
import com.gregtechceu.gtceu.api.machine.trait.NotifiableFluidTank
import com.lowdragmc.lowdraglib.side.fluid.FluidStack
import com.lowdragmc.lowdraglib.side.fluid.IFluidTransfer
import com.lowdragmc.lowdraglib.side.fluid.forge.FluidHelperImpl
import net.minecraft.core.BlockPos
import net.minecraft.core.Direction
import net.minecraft.world.level.Level
import net.minecraftforge.common.capabilities.ForgeCapabilities
import net.minecraftforge.fluids.capability.IFluidHandler
import java.util.function.Predicate

open class SuperNotifiableFluidTank(
    machine: MetaMachine,
    slots: Int,
    capacity: Long,
    io: IO
) : NotifiableFluidTank(machine, slots, capacity, io) {

    override fun exportToNearby(vararg facings: Direction) {
        if (!isEmpty) {
            machine.level?.let { level ->
                for (facing in facings) {
                    exportToTarget(
                        this,
                        machine.getFluidCapFilter(facing),
                        level,
                        machine.pos.relative(facing),
                        facing.opposite
                    )
                }
            }
        }
    }

    companion object {
        fun exportToTarget(
            source: IFluidTransfer,
            filter: Predicate<FluidStack>,
            level: Level,
            pos: BlockPos,
            direction: Direction?
        ) {
            val state = level.getBlockState(pos)
            if (!state.hasBlockEntity()) return

            val blockEntity = level.getBlockEntity(pos) ?: return
            val cap = blockEntity.getCapability(ForgeCapabilities.FLUID_HANDLER, direction).resolve()
            if (!cap.isPresent) return

            val target = cap.get()
            for (srcIndex in 0 until source.tanks) {
                val currentFluid = source.getFluidInTank(srcIndex)
                if (currentFluid.isEmpty || !filter.test(currentFluid)) {
                    continue
                }

                val toDrain = currentFluid.copy()
                var remainAmount = currentFluid.amount
                do {
                    val filled = target.fill(
                        FluidHelperImpl.toFluidStack(source.drain(toDrain, true)),
                        IFluidHandler.FluidAction.SIMULATE
                    )
                    if (filled > 0) {
                        toDrain.amount = filled.toLong()
                        target.fill(
                            FluidHelperImpl.toFluidStack(source.drain(toDrain, false)),
                            IFluidHandler.FluidAction.EXECUTE
                        )
                        remainAmount -= filled
                    } else {
                        break
                    }
                    toDrain.amount = remainAmount
                } while (!toDrain.isEmpty)
            }
        }
    }
}